backend/src/models/User.ts

```diff
@@ -0,0 +1,328 @@
+import {Table, Column, Model} from 'sequelize-typescript';
+import * as util from 'util';
+import * as crypto from 'crypto';
+import * as _ from 'lodash';
+import {Strategy} from 'passport-local';
+
+const LocalStrategy = Strategy;
+
+// The default option values
+const defaultAttachOptions = {
+    selectFields: false,
+    activationkeylen:  8,
+    resetPasswordkeylen:  8,
+    saltlen:  32,
+    iterations:  12000,
+    keylen:  512,
+    digest:  'sha1',
+    usernameField: 'username',
+    usernameLowerCase: false,
+    activationRequired: false,
+    hashField: 'hash',
+    saltField: 'salt',
+    activationKeyField: 'activationKey',
+    resetPasswordKeyField: 'resetPasswordKey',
+    incorrectPasswordError: 'Incorrect password',
+    incorrectUsernameError: 'Incorrect username',
+    invalidActivationKeyError: 'Invalid activation key',
+    invalidResetPasswordKeyError: 'Invalid reset password key',
+    missingUsernameError: 'Field %s is not set',
+    missingFieldError: 'Field %s is not set',
+    missingPasswordError: 'Password argument not set!',
+    userExistsError: 'User already exists with %s',
+    activationError: 'Email activation required',
+    noSaltValueStoredError: 'Authentication not possible. No salt value stored in db!'
+};
+const options = defaultAttachOptions;
+
+/**
+ * Initially this table had the fields:
+ * - ID            BIGINT(20)
+ * - ADMIN         TINYINT(1)
+ * - CREATED       DATETIME
+ * - EMAIL         VARCHAR(255)
+ * - LASTMODIFIED  DATETIME
+ * - NAME          VARCHAR(255)
+ * - SLUG          VARCHAR(255)
+ * - CREATOR_ID    BIGINT(20)
+ * - LASTEDITOR_ID BIGINT(20)
+ */
+
+
+
+@Table({
+  timestamps: true
+})
+class User extends Model<User> {
+  @Column({
+    allowNull: false
+  })
+  email: string;
+
+  @Column({
+    allowNull: false
+  })
+  username: string;
+
+  @Column
+  hash: string;
+
+  @Column
+  salt: string;
+
+  @Column
+  activationKey: string;
+
+  @Column
+  resetPasswordKey: string;
+
+  @Column
+  verified: boolean;
+
+  setPassword (password, cb) {
+    if (!password) {
+      return cb(new Error(options.missingPasswordError));
+    }
+
+    const self = this;
+
+    crypto.randomBytes(options.saltlen, function (err, buf) {
+      if (err) {
+        return cb(err);
+      }
+
+      const salt = buf.toString('hex');
+
+      crypto.pbkdf2(password, salt, options.iterations, options.keylen, options.digest, function (err, hashRaw) {
+        if (err) {
+          return cb(err);
+        }
+
+        // TODO: type
+        self.set(options.hashField, new Buffer(hashRaw as any as string, 'binary').toString('hex'));
```

Comment:
The code wasn't really changed from the project I copied it from, I just added some types and stuff. This was the code he wrote and the type signature from Buffer didn't work for it. I don't want to touch it yet, because currently it's still the code from the project.